# Set directory for Kallisto output files (IMPORTANT: must be a directory containing only the directories with results of each sample)
# Transcript to gene map was generated by running script /data/users/jvaleriano/lncRNA/scripts/05_Transcript_to_Gene.sh
# The differential expression results file (paraclonal-parental) from Tièche et all was used as a quality check
KALLISTO_OUT_DIR <- "/Users/jazminvaleriano/Documents/lncRNA/kallisto_out"
TRANSCRIPT_GENE_MAP <- "/Users/jazminvaleriano/Documents/lncRNA/Metadata/transcript_to_gene.tsv"
TIECHE_RESULTS <- "/Users/jazminvaleriano/Documents/lncRNA/Tieche RNA-Seq DATA.csv"

## PREPARE ENVIRONMENT----------------------------------------------------------
# List of required packages from CRAN and bioconductor:
required_packages_cran <- c("devtools","ggplot2","ggrepel","dplyr","cowplot")
required_packages_bioc <- c("rhdf5","biomaRt","EnhancedVolcano")

# Function to check for packages and install from CRAN if missing
install_if_missing_CRAN <- function(package_name) {
  if (!requireNamespace(package_name, quietly = TRUE)) {
    install.packages(package_name)
  } else {
    message(paste("Package", package_name, "is already installed."))
  }
}

# Install from CRAN
invisible(sapply(required_packages_cran,install_if_missing_CRAN))

# Check for packages and install from Bioconductor v. 3.18
BiocManager::install(version = "3.18")
install_if_missing_bioc <- function(package_name){
  if (!requireNamespace(package_name,quietly = TRUE)){
    BiocManager::install(package_name)
  } else {
    message(paste("Package", package_name, "is already installed."))
  }
}
invisible(sapply(required_packages_bioc,install_if_missing_bioc))

if (!requireNamespace("sleuth",quietly=TRUE)){
  devtools::install_github("pachterlab/sleuth")
} else {
    message(paste("sleuth is already installed."))
  }

# Load all libraries
libraries <- c("sleuth",required_packages_cran,required_packages_bioc)
invisible(sapply(libraries, library, character.only = TRUE))

## PREPARE METADATA-------------------------------------------------------------
# Samples to conditions table
sample_id <- dir(KALLISTO_OUT_DIR)
kal_dirs <- file.path(KALLISTO_OUT_DIR, sample_id)
sampletype <- factor(c(rep("paraclonal", 3), rep("parental", 3)))

samples2conditions <-data.frame(sample=sample_id,condition=sampletype,path=kal_dirs)
samples2conditions$condition <- relevel(samples2conditions$condition, ref = "parental")

# Transcripts to genes table
# In order to include both the annotated and novel genes, we extracted the transcript id and gene id from gtf 
# meta assemble (output from stringTie) and merged it to the reference gtf from gencode 
ttg <- read.table(TRANSCRIPT_GENE_MAP, header = TRUE, sep = "\t", fill = TRUE)
colnames(ttg)[1] <- "target_id"

## CREATE SLEUTH OBJECT---------------------------------------------------------
# we provide Sleuth with the mapping to identify the transcripts that belong to each gene id. 
# we specify a transformation function to get log2fold change instead of n log. 
# to get results at the gene level, we use 'gene_id' as aggregation column

so <- sleuth_prep(samples2conditions, 
                  target_mapping = ttg, 
                  read_bootstrap_tpm = TRUE,
                  extra_bootstrap_summary = TRUE,
                  transformation_function = function(x) log2(x + 0.5),
                  aggregation_column = 'gene_id',)

# Exploratory samples PCA and Heatmap as a quality check. 
#sleuth_live(so)
plot_pca(so, text_labels= T , color_by = 'condition', units= "tpm")
pdf('sample_heatmap.pdf')
plot_sample_heatmap(so, use_filtered = TRUE, color_high = "white",
                    color_low = "dodgerblue", x_axis_angle = 50,
                    annotation_cols = setdiff(colnames(so$sample_to_covariates), "sample"),
                    cluster_bool = TRUE)
dev.off()

#Sleuth generates an abundance table in the SO with the sample names and conditions
transcript_abundance_table<-kallisto_table(so, use_filtered = FALSE, normalized = TRUE,
               include_covariates = TRUE)
write.csv(transcript_abundance_table, file = "Transcript_abundance_table.csv", row.names = FALSE)

# Fit the full model to  “smooth” the raw kallisto abundance estimates for each sample 
# using a linear model with a parameter that represents the experimental condition 
# (in this case parental vs paraclonal)
so <- sleuth_fit(so, ~condition, 'full')

# To test for transcripts that are differential expressed between the conditions, 
# sleuth performs a second fit to a “reduced” model that presumes abundances are 
# equal in the two conditions.
so <- sleuth_fit(so, ~1, 'reduced')

# Check that both models have been fit
#models(so)

# Likelihood ratio test so sleuth identifies genes whose abundances are significantly 
# better explained by the full model (When paraclonal condition is taken into account):
so <- sleuth_lrt(so, 'reduced', 'full')

# Examine the significant results (gene level)
lrt_results <- sleuth_results(so, 'reduced:full', 'lrt', show_all = FALSE)

#Filter for significant, at qval <= 0.05, and only including protein coding, novel and lncRNA genes
biotypes_of_interest <- c("", "protein_coding", "lncRNA")
alpha <- 0.05

lrt_significant_genes <- lrt_results %>%
  filter(qval <= alpha, biotype %in% biotypes_of_interest)

# Save significant results in a csv
write.csv(lrt_significant_genes, file = "lrt_significant_genes.csv", row.names = FALSE)

# Wald test for differential expression: This function computes the Wald test on 
# one specific ‘beta’ coefficient on every transcript; this offers beta values 
# which are biased folds changes values for pairwise comparisons

so <- sleuth_wt(so, which_beta = 'conditionparaclonal','full')

# output results
tr_wt_results <- sleuth_results(so, 'conditionparaclonal', test_type = 'wt', pval_aggregate = F)

# Change the name of the beta value columns
names(tr_wt_results)[names(tr_wt_results) == "b"] <- "log2fold_change"
names(tr_wt_results)[names(tr_wt_results) == "se_b"] <- "se_log2fold_change"

# Filter results for significant only, and biotypes of interest. 
wt_significant_transcripts <- tr_wt_results %>%
  filter(qval <= alpha, biotype %in% biotypes_of_interest)
write.csv(wt_significant_transcripts, file = "wt_significant_transcripts.csv", row.names = FALSE)

# QUALITY CHECK 
# Compare results from my analysis with the one obtained by Tièche et al. (only named genes)
tieche_results <- read.csv(TIECHE_RESULTS)

# Filter significant results, and dif. expressed at a log2FoldChange threshold of 2
tieche_significant_de <- tieche_results %>%
  filter(padj <= alpha, log2FoldChange <= -2 | log2FoldChange >= 2 )

my_significant_de_transcripts <- tr_wt_results %>%
  filter(qval <= alpha, log2fold_change <= -2 | log2fold_change >= 2 )

my_significant_de_genes <- lrt_results %>%
  filter(qval <= alpha)

Tieche_vs_my_transcripts <- intersect(tieche_significant_de$hgnc_symbol, 
                                      my_significant_de_transcripts$gene_name)
total_intersect_vs_transcripts <- length(Tieche_vs_my_transcripts)

Tieche_vs_my_genes <- intersect(tieche_significant_de$hgnc_symbol, 
                                      my_significant_de_genes$gene_name)
total_intersect_vs_genes <- length(Tieche_vs_my_genes)

total_de_genes_Tieche <- length(unique(tieche_significant_de$hgnc_symbol))

percent_intersect_genes <- round(100*(total_intersect_vs_genes / total_de_genes_Tieche),2)
percent_intersect_transcripts <- round(100*(total_intersect_vs_transcripts / total_de_genes_Tieche),2)

## VOLCANO PLOTS ---------------------------------------------------------------

#Subset for lncRNA, protein coding and novel genes from the wt results (Transcript level)

lncRNA_wt_transcript <- subset(tr_wt_results, biotype == "lncRNA")
write.csv(lncRNA_wt_transcript, file = "lncRNA_wt_transcript.csv", row.names = FALSE)

protein_coding_wt_transcript <- subset(tr_wt_results, biotype == "protein_coding")
write.csv(protein_coding_wt_transcript, file = "protein_coding_wt_transcript.csv", row.names = FALSE)

novel_wt_transcript <- subset(tr_wt_results, gene_name=="")
write.csv(novel_wt_transcript, file = "novel_wt_transcript.csv", row.names= FALSE)


library(EnhancedVolcano)

lncRNA_volcano <-EnhancedVolcano(lncRNA_wt_transcript, x="log2fold_change", y="qval", 
                pCutoff = alpha, FCcutoff = 2,
                title = "lncRNA differential expression parental vs paraclonal",
                lab=lncRNA_wt_transcript$gene_name, labSize = 6,
                drawConnectors = T, 
                legendPosition = 'right', legendLabSize = 12,
)
ggsave("lncRNA_volcano_plot.png",lncRNA_volcano, width = 10, height = 8, units = "in")

protein_coding_volcano<-EnhancedVolcano(protein_coding_wt_transcript, x="log2fold_change", y="qval", 
                pCutoff = alpha, FCcutoff = 2,
                title = "Protein Coding genes differential expression parental vs paraclonal",
                lab=protein_coding_wt_transcript$gene_name, labSize = 6,
                drawConnectors = T, 
                legendPosition = 'right', legendLabSize = 12,
)
ggsave("protein_coding_volcano.png",protein_coding_volcano, width = 10, height = 8, units = "in")

novel_volcano <-EnhancedVolcano(novel_wt_transcript, x="log2fold_change", y="qval", 
                pCutoff = alpha , FCcutoff = 2,
                title = " Novel genes differential expression parental vs paraclonal",
                lab=novel_wt_transcript$gene_id, labSize = 6,
                drawConnectors = T, 
                legendPosition = 'right', legendLabSize = 12,
)
ggsave("novel_volcano.png",novel_volcano, width = 10, height = 8, units = "in")


#Heat Maps--------------------

# target_id <- "ENST00000257555.11"  # Replace with the actual transcript or gene identifier
# units <- "est_counts"
# color_by <- "condition"  # Replace with the actual grouping variable in your sample metadata
# x_axis_angle <- 50
# divide_groups <- TRUE
# 
# # Call the plot_bootstrap function
# plot_bootstrap(oe_transcript_level, target_id, units = units, color_by = color_by,
#                x_axis_angle = x_axis_angle, divide_groups = divide_groups)
# 
