# Set directory for Kallisto output files (IMPORTANT: must be a directory containing only the directories with 
# the Kallisto results of each sample)
# Transcript to gene map was generated by running /data/users/jvaleriano/lncRNA/scripts/05_Transcript_to_Gene.sh
KALLISTO_OUT_DIR <- "/Users/jazminvaleriano/Documents/lncRNA/kallisto_out"
TRANSCRIPT_GENE_MAP <- "/Users/jazminvaleriano/Documents/lncRNA/Metadata/transcript_to_gene.tsv"

## Prepare environment---------------------------------------------------------
# List of required packages from CRAN and bioconductor:
required_packages_cran <- c("devtools","ggplot2","ggrepel","dplyr","pheatmap")
required_packages_bioc <- c("rhdf5","biomaRt","EnhancedVolcano")

# Function to check for packages and install from CRAN if missing
install_if_missing_CRAN <- function(package_name) {
  if (!requireNamespace(package_name, quietly = TRUE)) {
    install.packages(package_name)
  } else {
    message(paste("Package", package_name, "is already installed."))
  }
}

# Install from CRAN
invisible(sapply(required_packages_cran,install_if_missing_CRAN))

# Check for packages and install from Bioconductor v. 3.18
BiocManager::install(version = "3.18")
install_if_missing_bioc <- function(package_name){
  if (!requireNamespace(package_name,quietly = TRUE)){
    BiocManager::install(package_name)
  } else {
    message(paste("Package", package_name, "is already installed."))
  }
}
invisible(sapply(required_packages_bioc,install_if_missing_bioc))

if (!requireNamespace("sleuth",quietly=TRUE)){
  devtools::install_github("pachterlab/sleuth")
} else {
    message(paste("sleuth is already installed."))
  }

# Load all libraries
libraries <- c("sleuth",required_packages_cran,required_packages_bioc)
invisible(sapply(libraries, library, character.only = TRUE))

## Prepare the metadata-----------------------------------
# Samples to conditions table
sample_id <- dir(KALLISTO_OUT_DIR)
kal_dirs <- file.path(KALLISTO_OUT_DIR, sample_id)
sampletype <- factor(c(rep("paraclonal", 3), rep("parental", 3)))

samples2conditions <-data.frame(sample=sample_id,condition=sampletype,path=kal_dirs)
samples2conditions$condition <- relevel(samples2conditions$condition, ref = "parental")

# Transcripts to genes table
# In order to include our novel genes, we extracted the transcript id and gene id from gtf 
# meta assemble (output from stringTie) and merged it to the reference gtf from gencode 
ttg <- read.table(TRANSCRIPT_GENE_MAP, header = TRUE, sep = "\t", fill = TRUE)
colnames(ttg)[1] <- "target_id"

## Run analysis at the transcript level-----------------------
# We first run Sleuth in transcript mode, we provide Sleuth with the mapping to 
# identify the transcripts that belong to each gene id. 
# we specify a transformation function to get log2fold change instead of n log. 
so_t <- sleuth_prep(samples2conditions, 
                  target_mapping = ttg, 
                  read_bootstrap_tpm = TRUE,
                  extra_bootstrap_summary = TRUE,
                  transformation_function = function(x) log2(x + 0.5))


# Fit the full model to  “smooth” the raw kallisto abundance estimates for each sample 
# using a linear model with a parameter that represents the experimental condition 
# (in this case parental vs paraclonal)
so_t <- sleuth_fit(so_t, ~condition, 'full')

# To test for transcripts that are differential expressed between the conditions, 
# sleuth performs a second fit to a “reduced” model that presumes abundances are 
# equal in the two conditions.
so_t <- sleuth_fit(so_t, ~1, 'reduced')

# Check that both models have been fit
# models(so)

# Likelihood ratio test so sleuth identifies transcripts whose abundances are significantly 
# better explained by the full model (When paraclonal condition is taken into account):
so_t <- sleuth_lrt(so_t, 'reduced', 'full')

# Exploratory PCA
plot_pca(so_t, text_labels= T , color_by = 'condition', units= "tpm")

# Examine the significant results
tr_lrt_results <- sleuth_results(so_t, 'reduced:full', 'lrt', show_all = FALSE)
tr_lrt_significant <- dplyr::filter(tr_lrt_results, qval <= 0.01)

# Save significant results in a csv
write.csv(tr_lrt_significant, file = "tr_lrt_significant.csv", row.names = FALSE)

# Wald test for differential expression: This function computes the Wald test on 
# one specific ‘beta’ coefficient on every transcript; this offers beta values 
# which are biased folds changes values for pairwise comparisons

so_t <- sleuth_wt(so_t, which_beta = 'conditionparaclonal','full')

# output results
tr_wt_results <- sleuth_results(so_t, 'conditionparaclonal', test_type = 'wt')

# Change the name of the beta value columns
names(tr_wt_results)[names(tr_wt_results) == "b"] <- "log2fold_change"
names(tr_wt_results)[names(tr_wt_results) == "se_b"] <- "se_log2fold_change"

tr_wt_significant <- dplyr::filter(tr_wt_results, qval <= 0.05)
write.csv(tr_wt_significant, file = "tr_wt_significant.csv", row.names = FALSE)

#Subset for lncRNA, protein coding and novel genes 
#lncRNA
lncRNA_wt_transcript <- subset(tr_wt_results, biotype == "lncRNA")
write.csv(lncRNA_wt_transcript, file = "lncRNA_wt_transcript.csv", row.names = FALSE)

protein_coding_wt_transcript <- subset(tr_wt_results, biotype == "protein_coding")
write.csv(protein_coding_wt_transcript, file = "protein_coding_wt_transcript.csv", row.names = FALSE)

lncRNA_and_coding_tr <- rbind(lncRNA_wt_transcript, protein_coding_wt_transcript)
write.csv(lncRNA_and_coding_tr, file = "lncRNA_and_coding_transcript.csv", row.names = FALSE)

novel_wt_transcript <- subset(tr_wt_results, gene_name=="")
write.csv(novel_wt_transcript, file = "novel_wt_transcript.csv", row.names= FALSE)

top20_novel_tr <- head(novel_wt_transcript,20)
write.csv(top20_novel_tr,file="top20_novel_tr.csv", row.names = FALSE)


## Run analysis at the gene level-----------------------
# Now we run Sleuth in gene mode, therefore we use 'gene_id' as aggregation column
# and provide Sleuth with the mapping to identify the transcripts that belong to each gene id. 
# we specify a transformation function to get log2fold change instead of n log. 
so_g <- sleuth_prep(samples2conditions, 
                  ~ condition,
                  target_mapping = ttg, 
                  aggregation_column = 'gene_id',
                  extra_bootstrap_summary = TRUE)


# Fit the models
so_g <- sleuth_fit(so_g, ~condition, 'full')
so_g <- sleuth_fit(so_g, ~1, 'reduced')

# Check that both models have been fit
models(so_g)

# Likelihood ratio test so sleuth identifies genes whose abundances are significantly 
# better explained by the full model (When paraclonal condition is taken into account):
so_g <- sleuth_lrt(so_g, 'reduced', 'full')
ge_lrt_results <- sleuth_results(so_g, 'reduced:full', 'lrt', show_all = FALSE)
ge_lrt_significant <- dplyr::filter(ge_lrt_results, qval <= 0.05)

# Wald test for differential expression:

so_g <- sleuth_wt(so_g, which_beta = 'conditionparaclonal','full')
ge_wt_results <- sleuth_results(so_g, 'conditionparaclonal', test_type = 'wt')

# Change the name of the beta value columns
#names(ge_wt_results)[names(ge_wt_results) == "b"] <- "log2fold_change"
#names(ge_wt_results)[names(ge_wt_results) == "se_b"] <- "se_log2fold_change"
ge_wt_significant <- dplyr::filter(ge_wt_results, qval <= 0.05)

# Save significant results in a csv
write.csv(ge_lrt_significant, file = "ge_lrt_significant.csv", row.names = FALSE)

#Subset wt results for lncRNA, protein coding and novel genes 
#lncRNA
lncRNA_wt_genelevel <- subset(ge_wt_results, biotype == "lncRNA")
write.csv(lncRNA_wt_genelevel, file = "lncRNA_wt_genelevel.csv", row.names = FALSE)

protein_coding_wt_genelevel <- subset(ge_wt_results, biotype == "protein_coding")
write.csv(protein_coding_wt_genelevel, file = "protein_coding_wt_genelevel.csv", row.names = FALSE)

novel_wt_genelevel <- subset(ge_wt_results, gene_name=="")
write.csv(novel_wt_genelevel, file = "novel_wt_genelevel.csv", row.names= FALSE)

top20_novel_genelevel <- head(novel_wt_genelevel,20)
write.csv(top20_novel_tr,file="top20_novel_tr.csv", row.names = FALSE)


#Volcano plots--------------------


library(EnhancedVolcano)

EnhancedVolcano(lncRNA_wt_transcript, x="log2fold_change", y="qval", 
                pCutoff = 1e-3, FCcutoff = 2,
                title = "lncRNA differential expression parental vs paraclonal",
                lab=lncRNA_wt_transcript$gene_name, labSize = 6,
                drawConnectors = T, 
                legendPosition = 'right', legendLabSize = 12,
)

EnhancedVolcano(protein_coding_wt_transcript, x="log2fold_change", y="qval", 
                pCutoff = 1e-3, FCcutoff = 2,
                title = "Protein Coding genes differential expression parental vs paraclonal",
                lab=protein_coding_wt_transcript$gene_name, labSize = 6,
                drawConnectors = T, 
                legendPosition = 'right', legendLabSize = 12,
)

EnhancedVolcano(novel_wt_transcript, x="log2fold_change", y="qval", 
                pCutoff = 1e-3, FCcutoff = 2,
                title = " Novel genes differential expression parental vs paraclonal",
                lab=novel_wt_transcript$gene_id, labSize = 6,
                drawConnectors = T, 
                legendPosition = 'right', legendLabSize = 12,
)
