# Set directory for Kallisto output files (IMPORTANT: must be a directory containing only the directories with results of each sample)
# Transcript to gene map was generated by running script /data/users/jvaleriano/lncRNA/scripts/05_Transcript_to_Gene.sh
# The differential expression results file (paraclonal-parental) from Tièche et all was used as a quality check
KALLISTO_OUT_DIR <- "/Users/jazminvaleriano/Documents/lncRNA/kallisto_out"
TRANSCRIPT_GENE_MAP <- "/Users/jazminvaleriano/Documents/lncRNA/Metadata/transcript_to_gene.tsv"
TIECHE_RESULTS <- "/Users/jazminvaleriano/Documents/lncRNA/Tieche RNA-Seq DATA.csv"

#Set significance value for statistic tests
alpha <- 0.01


## PREPARE ENVIRONMENT----------------------------------------------------------
# List of required packages from CRAN and bioconductor:
required_packages_cran <- c("devtools","ggplot2","ggrepel","dplyr","cowplot","VennDiagram")
required_packages_bioc <- c("rhdf5","biomaRt","EnhancedVolcano")

# Function to check for packages and install from CRAN if missing
install_if_missing_CRAN <- function(package_name) {
  if (!requireNamespace(package_name, quietly = TRUE)) {
    install.packages(package_name)
  } else {
    message(paste("Package", package_name, "is already installed."))
  }
}

# Install from CRAN
invisible(sapply(required_packages_cran,install_if_missing_CRAN))

# Check for packages and install from Bioconductor v. 3.18
BiocManager::install(version = "3.18")
install_if_missing_bioc <- function(package_name){
  if (!requireNamespace(package_name,quietly = TRUE)){
    BiocManager::install(package_name)
  } else {
    message(paste("Package", package_name, "is already installed."))
  }
}
invisible(sapply(required_packages_bioc,install_if_missing_bioc))

if (!requireNamespace("sleuth",quietly=TRUE)){
  devtools::install_github("pachterlab/sleuth")
} else {
    message(paste("sleuth is already installed."))
  }

# Load all libraries
libraries <- c("sleuth",required_packages_cran,required_packages_bioc)
invisible(sapply(libraries, library, character.only = TRUE))

## PREPARE METADATA-------------------------------------------------------------
# Samples to conditions table
sample_id <- dir(KALLISTO_OUT_DIR)
kal_dirs <- file.path(KALLISTO_OUT_DIR, sample_id)
sampletype <- factor(c(rep("paraclonal", 3), rep("parental", 3)))

samples2conditions <-data.frame(sample=sample_id,condition=sampletype,path=kal_dirs)
samples2conditions$condition <- relevel(samples2conditions$condition, ref = "parental")

# Transcripts to genes table
# In order to include both the annotated and novel genes, we extracted the transcript id and gene id from gtf 
# meta assemble (output from stringTie) and merged it to the reference gtf from gencode 
ttg <- read.table(TRANSCRIPT_GENE_MAP, header = TRUE, sep = "\t", fill = TRUE)
colnames(ttg)[1] <- "target_id"

## CREATE SLEUTH OBJECT---------------------------------------------------------
# we provide Sleuth with the mapping to identify the transcripts that belong to each gene id. 
# we specify a transformation function to get log2fold change instead of n log. 
# to get results at the gene level, we use 'gene_id' as aggregation column

so <- sleuth_prep(samples2conditions, 
                  target_mapping = ttg, 
                  read_bootstrap_tpm = TRUE,
                  extra_bootstrap_summary = TRUE,
                  transformation_function = function(x) log2(x + 0.5),
                  aggregation_column = 'gene_id',)

## DATA EXPLORATION ------------------------------------------------------------

# Exploratory samples PCA and Heatmap as a quality check. 
#sleuth_live(so)
plot_pca(so, text_labels= T , color_by = 'condition', units= "tpm")
ggsave('pca.sleuth.png')

pdf('sample_heatmap.pdf')
plot_sample_heatmap(so, use_filtered = TRUE, color_high = "white",
                    color_low = "dodgerblue", x_axis_angle = 50,
                    annotation_cols = setdiff(colnames(so$sample_to_covariates), "sample"),
                    cluster_bool = TRUE)
dev.off()

#Sleuth generates an abundance table in the SO with the sample names and conditions
transcript_abundance_table<-kallisto_table(so, use_filtered = FALSE, normalized = TRUE,
               include_covariates = TRUE)
write.csv(transcript_abundance_table, file = "Transcript_abundance_table.csv", row.names = FALSE)

## MODEL FIT AND DIFF EXP ANALYSIS ---------------------------------------------
# Fit the full model to  “smooth” the raw kallisto abundance estimates for each sample 
# using a linear model with a parameter that represents the experimental condition 
# (in this case parental vs paraclonal)
so <- sleuth_fit(so, ~condition, 'full')

# To test for transcripts that are differential expressed between the conditions, 
# sleuth performs a second fit to a “reduced” model that presumes abundances are 
# equal in the two conditions.
so <- sleuth_fit(so, ~1, 'reduced')

# Check that both models have been fit
#models(so)

# Likelihood ratio test so sleuth identifies genes whose abundances are significantly 
# better explained by the full model (When paraclonal condition is taken into account):
so <- sleuth_lrt(so, 'reduced', 'full')

# Examine the significant results (gene level)
lrt_results <- sleuth_results(so, 'reduced:full', 'lrt', show_all = FALSE)

#Filter for significant
lrt_significant <- lrt_results %>%filter(qval <= alpha)
# Save significant results in a csv
write.csv(lrt_significant, file = "wt_significant.csv", row.names = FALSE)
#Count unique genes identified
total_de_genes_gl<-length(unique(lrt_significant$gene_name))

#Subset for lncRNA, novel and protein coding 

lncRNA_lrt_significant<- subset(lrt_significant, biotype == "lncRNA")
write.csv(lncRNA_lrt_significant, file = "lncRNA_lrt_significant.csv", row.names = FALSE)
lrt_num_significant_lncRNA <- length(unique(lncRNA_lrt_significant$gene_name))

protein_coding_lrt_significant<- subset(lrt_significant, biotype == "protein_coding")
write.csv(protein_coding_lrt_significant, file = "protein_coding_lrt_significant.csv", row.names = FALSE)
lrt_num_significant_pc <- length(unique(protein_coding_lrt_significant$gene_name))

novel_lrt_significant<- subset(lrt_significant, biotype=="novel")
write.csv(novel_lrt_significant, file = "novel_lrt_significant.csv", row.names= FALSE)
lrt_num_significant_novel <- length(unique(novel_lrt_significant$gene_name))

cat("Total DE genes at gene level LRT analysis:", total_de_genes_gl,"\n",
    "Total DE lncRNA at gene level LRT analysis:", lrt_num_significant_lncRNA,"\n",
    "Total DE protein coding at gene level LRT analysis:", lrt_num_significant_pc,"\n",
    "Total DE novel genes at gene level LRT analysis:", lrt_num_significant_novel)


# Wald test for differential expression: This function computes the Wald test on 
# one specific ‘beta’ coefficient on every transcript; this offers beta values 
# which are biased folds changes values for pairwise comparisons

so <- sleuth_wt(so, which_beta = 'conditionparaclonal','full')

# output results, by deactivating the option of pval_aggregate we go down to transcript level
tr_wt_results <- sleuth_results(so, 'conditionparaclonal', test_type = 'wt', pval_aggregate = F)

# Change the name of the beta value columns
names(tr_wt_results)[names(tr_wt_results) == "b"] <- "log2fold_change"
names(tr_wt_results)[names(tr_wt_results) == "se_b"] <- "se_log2fold_change"

# Filter results for significant 
wt_significant <- tr_wt_results %>%  filter(qval <= alpha)
write.csv(wt_significant, file = "wt_significant.csv", row.names = FALSE)
total_de_genes_tl<-length(unique(wt_significant$gene_name))


lncRNA_wt_significant<- subset(wt_significant, biotype == "lncRNA")
write.csv(lncRNA_wt_significant, file = "lncRNA_wt_significant.csv", row.names = FALSE)
wt_num_significant_lncRNA <- length(unique(lncRNA_wt_significant$gene_name))

protein_coding_wt_significant<- subset(wt_significant, biotype == "protein_coding")
write.csv(protein_coding_wt_significant, file = "protein_coding_lrt_significant.csv", row.names = FALSE)
wt_num_significant_pc <- length(unique(protein_coding_wt_significant$gene_name))

novel_wt_significant<- subset(wt_significant, biotype=="novel")
write.csv(novel_wt_significant, file = "novel_wt_significant.csv", row.names= FALSE)
wt_num_significant_novel <- length(unique(novel_wt_significant$gene_name))

cat("Total DE genes at transcript level WT analysis:", total_de_genes_tl,"\n",
    "Total DE lncRNA at gene level LRT analysis:", wt_num_significant_lncRNA,"\n",
    "Total DE protein coding at gene level LRT analysis:", wt_num_significant_pc,"\n",
    "Total DE novel genes at gene level LRT analysis:", wt_num_significant_novel)

## QUALITY CHECK---------------------------------------------------------------- 
# Compare results from my analysis with the one obtained by Tièche et al. (only named/annotated genes)
tieche_results <- read.csv(TIECHE_RESULTS)
tieche_results$hgnc_symbol <- ifelse(tieche_results$hgnc_symbol == "", 
                                     tieche_results$ensembl_gene_id,
                                     tieche_results$hgnc_symbol)


# Filter significant results. 
tieche_significant_de <- tieche_results %>%
  filter(padj <= alpha)

wt_significant_transcripts <- tr_wt_results %>%
  filter(qval <= alpha)

wt_significant <- lrt_results %>%
  filter(qval <= alpha)

#Get the list of known genes identified as diff expressed in each dataset and plot them in a Venn Diagram. 
known_genes_tieche <- unique(na.exclude(tieche_significant_de$hgnc_symbol))
known_genes_WT <- unique(na.exclude(wt_significant_transcripts$gene_name))
known_genes_lrt <- unique(na.exclude(wt_significant$gene_name))

venn.diagram(x=list(known_genes_tieche,known_genes_lrt,known_genes_WT),
             category.names = c("Tièche et al." , "WT - Transcript level " , "LRT - Gene level"),
             filename = 'venn_diagramm.png',
             output=TRUE,
             fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3)),
             )

#Venn diagram with only the top 20 most significant results from Tieche
top100_genes_tieche <- head(unique(na.exclude(tieche_significant_de$hgnc_symbol)),100)
venn.diagram(x=list(top100_genes_tieche,known_genes_WT,known_genes_lrt),
             category.names = c("Tièche et al." , "WT - Transcript level " , "LRT - Gene level"),
             filename = 'venn_diagramm_top.png',
             output=TRUE,
             fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3)),
)


known_genes_df<-data.frame(Tieche=numeric(1), Gene_level=numeric(1), Transcript_level=numeric(1))
known_genes_df$Tieche<-length(known_genes_tieche)
known_genes_df$Gene_level<-length(known_genes_lrt)
known_genes_df$Transcript_level<-length(known_genes_WT)

intersecting_genes<-data.frame(Tieche_vs_LRT=numeric(1),Tieche_vs_WT=numeric(1),LRT_vs_WT=numeric(1))
intersecting_genes$Tieche_vs_LRT <- length(intersect(known_genes_tieche,known_genes_lrt))
intersecting_genes$Tieche_vs_WT <- length(intersect(known_genes_tieche,known_genes_WT))
intersecting_genes$LRT_vs_WT <- length(intersect(known_genes_lrt,known_genes_WT))


#Percentage of genes intersecting with those found by Tièche et al at gene level (LRT) and transcript level (WT)
percent_genes_in_gl <- round(100*(intersecting_genes$Tieche_vs_LRT / known_genes_df$Tieche),2)
percent_genes_in_tl <- round(100*(intersecting_genes$Tieche_vs_WT / known_genes_df$Tieche),2)
percent_genes_in_gl
percent_genes_in_tl


## VOLCANO PLOTS ---------------------------------------------------------------

#Subset for lncRNA, protein coding and novel genes from the wt results (Transcript level)

lncRNA_wt_transcript <- subset(tr_wt_results, biotype == "lncRNA")
write.csv(lncRNA_wt_transcript, file = "lncRNA_wt_transcript.csv", row.names = FALSE)

protein_coding_wt_transcript <- subset(tr_wt_results, biotype == "protein_coding")
write.csv(protein_coding_wt_transcript, file = "protein_coding_wt_transcript.csv", row.names = FALSE)

novel_wt_transcript <- subset(tr_wt_results, biotype =="novel")
write.csv(novel_wt_transcript, file = "novel_wt_transcript.csv", row.names= FALSE)


library(EnhancedVolcano)

lncRNA_volcano <-EnhancedVolcano(lncRNA_wt_transcript, x="log2fold_change", y="qval", 
                pCutoff = alpha, FCcutoff = 2,
                title = "A",
                lab=lncRNA_wt_transcript$gene_name, labSize = 6,
                drawConnectors = T, 
                legendPosition = 'right', legendLabSize = 12,
)
ggsave("lncRNA_volcano_plot.png",lncRNA_volcano, width = 10, height = 8, units = "in")

protein_coding_volcano<-EnhancedVolcano(protein_coding_wt_transcript, x="log2fold_change", y="qval", 
                pCutoff = alpha, FCcutoff = 2,
                title = "B",
                lab=protein_coding_wt_transcript$gene_name, labSize = 6,
                drawConnectors = T, 
                legendPosition = 'right', legendLabSize = 12,
)
ggsave("protein_coding_volcano.png",protein_coding_volcano, width = 10, height = 8, units = "in")

novel_volcano <-EnhancedVolcano(novel_wt_transcript, x="log2fold_change", y="qval", 
                pCutoff = alpha , FCcutoff = 2,
                title = "C",
                lab=novel_wt_transcript$gene_id, labSize = 6,
                drawConnectors = T, 
                legendPosition = 'right', legendLabSize = 12,
)

ggsave("novel_volcano.png",novel_volcano, width = 10, height = 8, units = "in")

## BOOTSTRAP PLOT --------------------------------------------------------------
plot_bootstrap(so, target_id="MSTRG.19367.2", units = "tpm", x_axis_angle = 50, divide_groups = T)
#sleuth_live(so)


